### DISCUSSION TITLE
.NET Port

### DISCUSSION THREAD

**Discussion by AdamTheLawyer** [freelawproject/eyecite#131] [General]
Has anyone thought about porting the library to .net (VB/C#/etc) ... either through a direct port or with an intermediate schmear of Python?

It would be nice to be able to write custom MS Word extensions that poke at citations.

**Comment by AdamTheLawyer** [comment]
Seems to work just fine using Python.Included as an intermediate layer. 

If I end up writing anything useful, I'll be sure to share.

**Comment by mlissner** [comment]
Fun. Porting comes up every now and then. I think somebody wanted to Rust the other day? One thing to note is that if you want performance, the HyperscanTokenizer is written in C. Not sure how that plays with a python..."schmear?"

**Comment by AdamTheLawyer** [comment]
Accessing the library from .NET languages isn't too difficult ... I used a translation layer called Python.Included that seems to work fine.

Now I'm trying to find a good way to use EyeCite to identify and tag citations in a MS Word document. Dumping it into plaintext and searching the cites works easily enough, but it's more of a challenge to push the cites back into the original. Editing the underlying wordML document won't work...the citations are usually broken up by formatting tags a few times.

Maybe locate the citations with the plaintext, then use MS Word's search-and-replace functions to re-locate them and put tags in?

**Reply by flooie** [reply]
do you have an example file?

**Reply by AdamTheLawyer** [reply]
Of accessing the Python library via .NET? Kind of. Getting it to work was a little tricky.

I used a library called [Python.Included](https://www.nuget.org/packages/Python.Included/) that basically embeds its own mini Python installation in your app. It actually seems to work per machine rather than per app; by default it goes to:

%USERPROFILE%\AppData\Local\python-3.7.3-embed-amd64

there's a function to move the install folder, but I kept having problems with it.

According to the docs, you just run the methods below to set up your Python install and add eyecite. I had issues with it, though; the TryInstallPip() command would just hang. If I went into the mini-python directory it makes, though, I could run the "Lib/get-pip.py" script from the console. (And I think I probably installed eyecite that way, too).

It did work, though. I'm guessing my issues with install directories and hanging are connected to some environment variable or other.




Imports System.Threading.Tasks
Imports Python.Included
Imports Python.Runtime

Public Class Form1
	Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load

		Dim EyeCite As Object
		Dim Output As Object

		'Debug.Print(Python.Deployment.Installer.EmbeddedPythonHome)
		'Installer.SetupPython()
		' this usually just hung for me; i had to install pip manually
		'Installer.TryInstallPip()

		' I think i did this manually too
		'Installer.PipInstallModule("eyecite")

		PythonEngine.Initialize()

		EyeCite = Py.Import("eyecite")

		Output = EyeCite.get_citations("Hi there. 100 F.3d 200. Farts")
		Debug.Print(Output.count)

	End Sub
End Class
```

**Comment by mlissner** [comment]
Very cool! I ran into the mixed tags problem a few times over the years. It's hard. Eyecite has some heuristics it uses to deal with that in HTML that you could check out, but yeah, that's a tough one. At my old job the engineers just gave up.

**Comment by AdamTheLawyer** [comment]
I was mostly surprised that Word hadn't written this in ages ago...and that there's no simple, low cost solution to just tag the citations for you.
